### ISSUE TITLE
Drag and Drop not working using winappdriver

### ISSUE THREAD

**Issue by yatheesh1989** [microsoft/WinAppDriver#1223]
I have a scenario to copy paste the content from 1 grid cell to another grid cell in table grid by drag and drop.
I have tried Action class with element and action with offset and Mouse Move/Down/Up nothing working
Its clicking and the source element after click the mouse move or drag is not happening
Using Mouse Move:
session.Mouse.MouseMove(sorce.Coordinates);
session.Mouse.MouseDown(null);
session.Mouse.MouseMove(dest.Coordinates);
session.Mouse.MouseUp(null);

Using Action Class:
var a = dest.Coordinates;
var c = a.LocationInDom;
int x = c.X;
int y = c.Y;
Actions action = new Actions(session);
action.DragAndDropToOffset(source, y,x).Perform();

Actions action = new Actions(session);
action.MoveToElement(source).Perform();
action.DragAndDrop(source, dest).Build().Perform();

Actions action = new Actions(session);
action.MoveToElement(source).Build().Perform();
action.ClickAndHold(source).Build().Perform();
action.MoveToElement(dest).Build().Perform();
action.MoveToElement(dest).Click().Build().Perform();

I have tried different way but all the time its clicking and holding the element to drag

**Comment by anunay1** [comment]
Have you tried with clipboard class

**Comment by yatheesh1989** [comment]
> Have you tried with clipboard class

@anunay1  no i didnt tried that. if you have any reference can you share here?

And the issue https://github.com/Microsoft/WinAppDriver/issues/219 mentioned as fixed but same i'm trying but not working

**Comment by anunay1** [comment]
import 
using System.Windows.Forms;

Clipboard.SetText(driver.Findelementbyname("").Text);
driver.findElementbyname(destination).click
driver.sendkeys(Keys.Control + "v")

**Comment by yatheesh1989** [comment]
> import
> using System.Windows.Forms;
> 
> Clipboard.SetText(driver.Findelementbyname("").Text);
> driver.findElementbyname(destination).click
> driver.sendkeys(Keys.Control + "v")

Is Clipboard working as copy paste right ? but here i want drag the one box element into the another box  its not a text. how can i do that with Clipboard? 
i have achieved the driver.sendkeys(Keys.Control + "C") and driver.sendkeys(Keys.Control + "v") without using clipboard like normal sendkeys() function

**Comment by anunay1** [comment]
Which version of windows application driver are u using. I am using 4.1.2, and its working perfectly at my end, just tried it now.

**Comment by anunay1** [comment]
> > import
> > using System.Windows.Forms;
> > Clipboard.SetText(driver.Findelementbyname("").Text);
> > driver.findElementbyname(destination).click
> > driver.sendkeys(Keys.Control + "v")
> 
> Is Clipboard working as copy paste right ? but here i want drag the one box element into the another box its not a text. how can i do that with Clipboard?
> i have achieved the driver.sendkeys(Keys.Control + "C") and driver.sendkeys(Keys.Control + "v") without using clipboard like normal sendkeys() function
> 
> And im not able to find the exact X and Y coordinates for the element in grid it giving X and Y but not in the expected location
> 
> How to find exact location of the element i tried with element.location and element.coordinates but it moving mouse some where . Can you please help me on this too?

If it is not able to find the exact coordinates you need to check first what coordinates it gives you, and then you can set offset values to point the mouse where u want.

**Comment by yatheesh1989** [comment]
> > > import
> > > using System.Windows.Forms;
> > > Clipboard.SetText(driver.Findelementbyname("").Text);
> > > driver.findElementbyname(destination).click
> > > driver.sendkeys(Keys.Control + "v")
> > 
> > 
> > Is Clipboard working as copy paste right ? but here i want drag the one box element into the another box its not a text. how can i do that with Clipboard?
> > i have achieved the driver.sendkeys(Keys.Control + "C") and driver.sendkeys(Keys.Control + "v") without using clipboard like normal sendkeys() function
> > And im not able to find the exact X and Y coordinates for the element in grid it giving X and Y but not in the expected location
> > How to find exact location of the element i tried with element.location and element.coordinates but it moving mouse some where . Can you please help me on this too?
> 
> If it is not able to find the exact coordinates you need to check first what coordinates it gives you, and then you can set offset values to point the mouse where u want.

i have WinAppDriver v1.2 Release Candidate from the link https://github.com/Microsoft/WinAppDriver/releases

Here im not seeing version 4.1.2. can you give me download link

**Comment by anunay1** [comment]
Sorry I have the same release, v1.2

**Comment by anunay1** [comment]
> > > import
> > > using System.Windows.Forms;
> > > Clipboard.SetText(driver.Findelementbyname("").Text);
> > > driver.findElementbyname(destination).click
> > > driver.sendkeys(Keys.Control + "v")
> > 
> > 
> > Is Clipboard working as copy paste right ? but here i want drag the one box element into the another box its not a text. how can i do that with Clipboard?
> > i have achieved the driver.sendkeys(Keys.Control + "C") and driver.sendkeys(Keys.Control + "v") without using clipboard like normal sendkeys() function
> > And im not able to find the exact X and Y coordinates for the element in grid it giving X and Y but not in the expected location
> > How to find exact location of the element i tried with element.location and element.coordinates but it moving mouse some where . Can you please help me on this too?
> 
> If it is not able to find the exact coordinates you need to check first what coordinates it gives you, and then you can set offset values to point the mouse where u want.

Can you try this?

**Comment by yatheesh1989** [comment]
> > > > import
> > > > using System.Windows.Forms;
> > > > Clipboard.SetText(driver.Findelementbyname("").Text);
> > > > driver.findElementbyname(destination).click
> > > > driver.sendkeys(Keys.Control + "v")
> > > 
> > > 
> > > Is Clipboard working as copy paste right ? but here i want drag the one box element into the another box its not a text. how can i do that with Clipboard?
> > > i have achieved the driver.sendkeys(Keys.Control + "C") and driver.sendkeys(Keys.Control + "v") without using clipboard like normal sendkeys() function
> > > And im not able to find the exact X and Y coordinates for the element in grid it giving X and Y but not in the expected location
> > > How to find exact location of the element i tried with element.location and element.coordinates but it moving mouse some where . Can you please help me on this too?
> > 
> > 
> > If it is not able to find the exact coordinates you need to check first what coordinates it gives you, and then you can set offset values to point the mouse where u want.
> 
> Can you try this?

Yes. I Will try what you mentioned. But can you give me any other solution to achieve this?

**Comment by yatheesh1989** [comment]
I able to find the offset for particular element but here im working with table grid so i source and destination element will any one of the row and column here the calculation were different every time. So the calculating offset and doing drag and drop will be more difficult here and there are more no.of chances of failure. Why im not able achieve with Action class simple drag and drop function?

**Comment by anunay1** [comment]
@yatheesh1989 : The actions class has an existing issue with drag and drop on WIN10 machines. Check the below

https://github.com/SeleniumHQ/selenium/issues/7744

I don't understand why the row and columns have different dimension, if the app in under your control please ask the developer to fix this. 

Also will it be possible for you to share the screenshot of your table. if it does not violate privacy.

**Comment by yatheesh1989** [comment]
> @yatheesh1989 : The actions class has an existing issue with drag and drop on WIN10 machines. Check the below
> 
> [SeleniumHQ/selenium#7744](https://github.com/SeleniumHQ/selenium/issues/7744)
> 
> I don't understand why the row and columns have different dimension, if the app in under your control please ask the developer to fix this.
> 
> Also will it be possible for you to share the screenshot of your table. if it does not violate privacy.

    
Hi Please find the screenshot of table grid of application and the highlited cell is have an element that to drag and drop from source cell to destination cell
but coordinates are not the exact location which return by WinAppDriver.
The application is a WPF Grid.. The Grid Cells do not have predefined rows and columns and they are not static..They are using Virtualization to create the cells based on the data the Grid loads
The cells and rows appear as if they are always there. The visual elements are only created when they are in view.
Source Element:
    From Driver: Location    {X = 576 Y = 241}    System.Drawing.Point
    From Inspect.exe:
        Mouse move (842,271)
        ClickablePoint:    {x:846 y:261} 


Dest Element
    From Driver: Location    {X = 1757 Y = 241}    System.Drawing.Point
    From Inspect.exe:
        Mouse move (1710,263)
        ClickablePoint:    {x:1700 y:261}


As of now i found some work around to work with Drag and Drop in data grid.

Actions action = new Actions(session);
action.MoveToElement(source).Build().Perform();
action.ClickAndHold(source).MoveByOffset(dest.Location.X, dest.Location.Y).Build().Perform();
dest.Click();
action.Release().Perform();

**Comment by Shakevg** [comment]
@yatheesh1989 Greate solution, working for me too
